### DISCUSSION TITLE
redirect multiple dynamic pages to root page (pages/blog/:id, pages/categories/:id --> pages/:id)

### DISCUSSION THREAD

**Discussion by tafelito** [vercel/next.js#13106] [Help]
Hi!

Anybody know if is possible with either NextJS or with Now to do a redirect from something like `pages/blog/:id` to `pages/:id` but for multiple pages, and by multiple pages I mean having `pages/blog/:id`, `pages/categories/:id` and then redirect all to the root `pages/:id`

I did this by creating a [slug].js inside my pages and then there dynamically check which component to show depending on the slug but the problem with that approach is that there is no way to respond with a 404 when a page is not found when using `fallback: true`

This is a follow up of https://github.com/zeit/next.js/discussions/10960

**Comment by sambhav-gore** [comment]
I would like to know this too. but for a subdomain route.
at the moment what I am doing is that I am checking for the url from req in `getInitialProps` and 302-redirecting for certain cases.

**Comment by tafelito** [comment]
#11646 
#15672

**Reply by AubreyHewes** [reply]
@tafelito The question is as described in referenced issues... and being currently handled in referenced issues..

see also https://github.com/vercel/next.js/issues/11646#issuecomment-669307405

You can add custom redirects in your config (seems to match what you are asking for)... https://nextjs.org/docs/api-reference/next.config.js/redirects#path-matching

**Reply by tafelito** [reply]
@AubreyHewes thanks, that's why I mentioned the reference issues so anyone can follow those. 
Even tho the recently added `fallback: 'blocking'` cover some uses cases, is does not cover mine. 
What I was most interested is what's actually coming soon, and is the way to properly handle the 404 when fallback is true and that's the second part of my question.
I did read the docs from the link you sent before, and that works great if a user enters to a `/blog/:id` or `/categories/:id` then I can redirect to `/:id` (it'd be even better if I could get the original request on the redirected page somehow), but if they go directly to `/:id` I still need to be able to handle that with the `[slug].js` page. 
And for that I need to be able to properly handle the 404 when `fallback: true`

**Reply by AubreyHewes** [reply]
Getting the original request on the redirected page will probably never be supported as the original request has been redirected (via the server) to another request..

yeah proper 404 are not handled yet when using `fallback: true`. this is also a problem for some of our projects; product catalogs which you do not want to generate on build, but on the fly. Not existing should be a proper 404 not a 200 with a page not found html output. I am waiting on that change ;-)

**Reply by tafelito** [reply]
I was reading the router docs I was wondering if something like this could be used for now before a proper way is actually supported 


```js
export default function Page({ post }) {
  const router = useRouter()

  if(router.isFallback) {
     return <h1>Loading...</h1>
  }

  useEffect(() => {
    router.beforePopState(({ url, as, options }) => {
      // I only want to allow these two routes!
      if (!post) {
        // Have SSR render bad routes as a 404.
        window.location.href = as
        return false
      }

      return true
    })
  }, [])

  return <h1>{post.title}</h1>
}
```

So basically on beforePopState if there is no post have the server render a 404
